pkgs/build-support/trivial-builders/default.nix

```diff
@@ -471,23 +475,36 @@ rec {
   symlinkJoin =
     args_@{ name
     , paths
+    , stripPrefix ? ""
     , preferLocalBuild ? true
     , allowSubstitutes ? false
     , postBuild ? ""
+    , failOnMissing ? stripPrefix == ""
     , ...
     }:
+    # Ensure no partial paths used, it will be confusing considering
+    # most of the time symlinkJoin is used with packages.
+    assert (stripPrefix == "" || (hasPrefix "/" stripPrefix) && stripPrefix != "/");
     let
-      args = removeAttrs args_ [ "name" "postBuild" ]
+      mapPaths = f: paths: map (path:
+        if path == null then null
+        else if isList path then mapPaths f path
+        else f path
+      ) paths;
+      args = removeAttrs args_ [ "name" "postBuild" "stripPrefix" "paths" "failOnMissing" ]
         // {
         inherit preferLocalBuild allowSubstitutes;
+        paths = mapPaths (path: "${path}${stripPrefix}") paths;
```

Comment:
Now that we're adding in something more than just a simple `for` loop, I think this needs a test. 

See the file `pkgs/build-support/trivial-builders/test/default.nix` for the set of tests.

Run the tests using `nix-build -A tests.trivial-builders.symlinkJoin`.

Take a look at `pkgs/test/replace-vars/default.nix` for an example of how to write those tests. `testers.testEqualContents` in particular will be useful, and `testBuildFailure` for demonstrating what happens when the derivation ought to error.

That'll give you the ability to do demonstrate the following cases:

1. Direct, normal use sans any path stripping (maybe with a nested list, per the recursive call).
2. Build failure due to a missing path.
3. Path stripping with missing paths
4. Path stripping with failures due to other reasons that are skipped.